cluster-autoscaler/cloudprovider/mcm/mcm_cloud_provider.go

```diff
@@ -146,7 +146,24 @@ func (mcm *mcmCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovider.N
 		return nil, nil
 	}
 
-	return mcm.mcmManager.GetMachineDeploymentForMachine(ref)
+	md, err := mcm.mcmManager.GetMachineDeploymentForMachine(ref)
```

Comment:
In our case we have MachineDeployments not managed by cluster-autoscaler. So it would find the machine but then cause an error because it's an unknown node group by CA (see the original error). Given the number of machine deployments is quite small I don't think this loop is a big deal. But we could stick it in a map possibly.

Reply:
@jsravn Looping over machine deployments seems to be sub-optimal here.

How about changing the following `GetMachineDeploymentForMachine` func to return `nil, nil` if there is no `MachineSet` found for the given `Machine`? I.e. enhance the code [here](https://github.com/gardener/autoscaler/blob/66334752174e00240db1bbc557a1acd165154bde/cluster-autoscaler/cloudprovider/mcm/mcm_manager.go#L352-L354) to additionally check for `kube_errors.IsNotFound()` like done [slightly above](https://github.com/gardener/autoscaler/blob/66334752174e00240db1bbc557a1acd165154bde/cluster-autoscaler/cloudprovider/mcm/mcm_manager.go#L336-L340) in the same function to the check for the `Machine` object.

Reply:
I still don't think this will help our usecase. Machine and MachineSet exist, but they are managed outside of cluster-autoscaler (via a manually configured MachineDeployment set to a static number of nodes). So the only way afaict is to compare against the list of cluster-autoscaler managed machine deployments.

Reply:
@amshuman-kr I modified it to use a map to avoid the loop, let me know what you think.